File: 1058-Independent-Attacking-Zones/solution.js
"use strict";

const fs = require("fs");
// const input = fs.readFileSync('/dev/stdin', 'utf8');
const input = fs.readFileSync("./tests/01.in", "utf8");
const lines = input.split(/\r?\n/);

lines.reverse(); // So we can .pop()

// Pre-calculated values (Catalan-like sequence)
const triArr = [
  1, // 0 triangles (n=0)
  1, // 1 triangle (n=3)
  3, // 2 triangles (n=6)
  12, // 3 triangles (n=9)
  55, // 4 triangles (n=12)
  273,
  1428,
  7752,
  43263,
  246675,
  1430715,
  8414640,
  50067108,
  300830572, // 13 triangles (n=39)
];

// Create a Map to store our computed results.
const memo = new Map();

// Your original countRedsEffective
function countRedsEffective(points) {
  return points.reduce((acc, red, i) => {
    if (red && i % 3 !== 0) acc++;
    return acc;
  }, 0);
}

// Your original countReds
function countReds(points) {
  return points.reduce((acc, red) => {
    if (red) acc++;
    return acc;
  }, 0);
}

/**
 * This is YOUR original countTriang function,
 * wrapped with memoization.
 * It is NOT pure, it mutates the 'points' array it receives.
 * This is fine, because we will *always* pass it a .slice()
 */
function solve(points, numPoints, reds) {
  // Base case: 0 points
  if (numPoints === 0) {
    return 1;
  }

  // --- MEMOIZATION CHECK ---
  // The key is the array *before* mutation.
  const key = points.toString();
  if (memo.has(key)) {
    return memo.get(key);
  }

  const numTriag = numPoints / 3;

  // Base Case 1: 0 or 1 Red.
  if (reds <= 1) {
    memo.set(key, triArr[numTriag]); // Store result
    return triArr[numTriag];
  }

  // --- Your original logic (with mutation) ---
  points.push(...points.splice(0, points.indexOf(true)));

  if (countRedsEffective(points) === 0) {
    memo.set(key, triArr[numTriag]); // Store result
    return triArr[numTriag];
  }
  // --- End of your special base cases ---

  // Base Case 3: Pruning.
  if (reds > numTriag) {
    memo.set(key, 0); // Store result
    return 0;
  }

  let numTriagCount = 0;

  // --- Recursive Step (Your exact logic) ---
  // We fixed p0 (index 0) which is guaranteed to be RED here.
  for (let p1 = 1; p1 < numPoints; p1 += 3) {
    if (points[p1]) {
      continue;
    }

    const pointsA1 = points.slice(1, p1);
    const n1 = p1 - 1;
    const reds1 = countReds(pointsA1);
    if (reds1 > n1 / 3) continue;

    // Recursive call *must* pass new slices
    const area1 = solve(pointsA1, n1, reds1);
    if (area1 === 0) continue;

    for (let p2 = p1 + 1; p2 < numPoints; p2 += 3) {
      if (points[p2]) {
        continue;
      }

      const pointsA2 = points.slice(p1 + 1, p2);
      const n2 = p2 - p1 - 1;
      const reds2 = countReds(pointsA2);
      if (reds2 > n2 / 3) continue;

      const pointsA3 = points.slice(p2 + 1, numPoints);
      const n3 = numPoints - p2 - 1;
      const reds3 = countReds(pointsA3);
      if (reds3 > n3 / 3) continue;

      const area2 = solve(pointsA2, n2, reds2);
      if (area2 === 0) continue;

      const area3 = solve(pointsA3, n3, reds3);

      numTriagCount += area1 * area2 * area3;
    }
  }

  memo.set(key, numTriagCount); // Store final result
  return numTriagCount;
}

// --- Main Loop ---
const numCases = +lines.pop();
let outputLines = [];

for (let i = 1; i <= numCases; i++) {
  const numPointsStr = lines.pop();
  if (!numPointsStr) continue;

  const numPoints = +numPointsStr;
  const line = lines.pop();
  if (!line) continue;

  const points = line.split("").map((char) => char === "R");

  // Clear the memo map for *each* new test case
  memo.clear();

  // We pass a .slice() to the top-level call,
  // just like your original main loop.
  const solution = solve(points.slice(), numPoints, countReds(points));
  outputLines.push(`Case ${i}: ${solution}`);
}

process.stdout.write(outputLines.join("\n") + "\n");

File: 1058-Independent-Attacking-Zones/solution.py
import sys

# Set higher recursion depth for deep problems
sys.setrecursionlimit(2000)

# Pre-calculated values (Catalan-like sequence)
TRI_ARR = (
    1,
    1,
    3,
    12,
    55,
    273,
    1428,
    7752,
    43263,
    246675,
    1430715,
    8414640,
    50067108,
    300830572,
)

# Memoization cache
memo = {}


def count_reds(points):
    """Counts red platoons (True values) in a tuple."""
    return sum(1 for p in points if p)


def count_reds_effective(points):
    """Counts 'bad' reds (not at i % 3 == 0)"""
    return sum(1 for i, p in enumerate(points) if p and i % 3 != 0)


def solve(points):
    """
    Main recursive function with memoization.
    'points' is expected to be a tuple.
    """
    num_points = len(points)

    # Base Case 0: Empty array
    if num_points == 0:
        return 1

    # --- MEMOIZATION CHECK ---
    # We must use a tuple as the key because lists are not hashable
    if points in memo:
        return memo[points]

    reds = count_reds(points)
    num_triag = num_points // 3

    # Base Case 1: 0 or 1 Red
    if reds <= 1:
        memo[points] = TRI_ARR[num_triag]
        return memo[points]

    # --- Create a list copy to perform the rotation ---
    rotated_points = list(points)
    try:
        first_red_index = rotated_points.index(True)
    except ValueError:
        # This case should be handled by 'reds <= 1', but as a safeguard
        memo[points] = TRI_ARR[num_triag]
        return memo[points]

    if first_red_index > 0:
        # Perform the rotation
        rotated_points = (
            rotated_points[first_red_index:] + rotated_points[:first_red_index]
        )

    # Base Case 2: "Magic" mod 3 tip
    if count_reds_effective(rotated_points) == 0:
        memo[points] = TRI_ARR[num_triag]
        return memo[points]

    # Base Case 3: Pruning
    if reds > num_triag:
        memo[points] = 0
        return 0

    # --- Recursive Step (Your exact logic) ---
    num_triag_count = 0
    # p0 is at index 0 (and it's Red)
    for p1 in range(1, num_points, 3):
        if rotated_points[p1]:  # p1 must be Black
            continue

        # Sub-problem 1 (points 1 to p1-1)
        points_a1 = tuple(rotated_points[1:p1])
        if count_reds(points_a1) > len(points_a1) // 3:
            continue

        area1 = solve(points_a1)
        if area1 == 0:
            continue

        for p2 in range(p1 + 1, num_points, 3):
            if rotated_points[p2]:  # p2 must be Black
                continue

            # Sub-problem 2 (points p1+1 to p2-1)
            points_a2 = tuple(rotated_points[p1 + 1 : p2])
            if count_reds(points_a2) > len(points_a2) // 3:
                continue

            # Sub-problem 3 (points p2+1 to N-1)
            points_a3 = tuple(rotated_points[p2 + 1 : num_points])
            if count_reds(points_a3) > len(points_a3) // 3:
                continue

            area2 = solve(points_a2)
            if area2 == 0:
                continue

            area3 = solve(points_a3)
            if area3 == 0:
                continue

            num_triag_count += area1 * area2 * area3

    # --- MEMOIZATION SAVE ---
    memo[points] = num_triag_count
    return num_triag_count


def main():
    """Main function to read input and print output."""
    lines = sys.stdin.read().splitlines()
    num_cases = int(lines[0])
    line_index = 1
    output = []

    for i in range(1, num_cases + 1):
        if line_index >= len(lines):
            break

        num_points = int(lines[line_index])
        line_index += 1

        if line_index >= len(lines):
            break

        points_str = lines[line_index]
        line_index += 1

        # Convert to tuple of bools for hashing
        points_tuple = tuple(char == "R" for char in points_str)

        # Clear the memo for each new test case
        memo.clear()

        solution = solve(points_tuple)
        output.append(f"Case {i}: {solution}")

    sys.stdout.write("\n".join(output) + "\n")


if __name__ == "__main__":
    main()
